CMakeLists.txt

```
@@ -0,0 +1,98 @@
+cmake_minimum_required (VERSION 3.8.2)
+project(imgui VERSION 1.73.0 LANGUAGES CXX C)
+
+option (IMGUI_EXAMPLES "Build ImGui examples" ON)
+option (IMGUI_DEMO "Include the ImGui demo window implementation in library" ON)
+
+option (IMGUI_IMPL_SDL "Build the SDL implementation (only if supported)" ON)
+option (IMGUI_IMPL_METAL "Build the Metal implementation (only if supported)" ${APPLE})
+option (IMGUI_IMPL_OSX "Build the OSX implementation (only if supported)" ${APPLE})
+option (IMGUI_IMPL_WIN32 "Build the Win32 (native winapi) implementation (only if supported)" ${WIN32})
+option (IMGUI_IMPL_GLFW "Build the GLFW implementation (only if supported)" ON)
+option (IMGUI_IMPL_GLUT "Build the GLUT implementation (only if supported)" ON)
+option (IMGUI_IMPL_OPENGL "Build the OpenGL implementation (only if supported)" ON)
+option (IMGUI_IMPL_OPENGL2 "Build the OpenGL 2 (legacy) implementation (only if supported)" ${IMGUI_IMPL_OPENGL})
+option (IMGUI_IMPL_DX9 "Build the DirectX 9 implementation (only if supported)" ${WIN32})
+option (IMGUI_IMPL_DX10 "Build the DirectX 10 implementation (only if supported)" ${WIN32})
+option (IMGUI_IMPL_DX11 "Build the DirectX 11 implementation (only if supported)" ${WIN32})
+option (IMGUI_IMPL_DX12 "Build the DirectX 12 implementation (only if supported)" ${WIN32})
+
+add_library (imgui
+    imgui.cpp
+    imgui_draw.cpp
+    imgui_widgets.cpp
+    $<$<BOOL:IMGUI_DEMO>:imgui_demo.cpp>
+)
+
+add_library(imgui::imgui ALIAS imgui)
+
+target_include_directories (imgui PUBLIC "${CMAKE_CURRENT_SOURCE_DIR}")
+
+target_compile_features(imgui PRIVATE cxx_std_98)
```

```suggestion
if (NOT CMAKE_CXX_STANDARD)
  target_compile_features(imgui PRIVATE cxx_std_98)
endif()
```

While it is not guaranteed that C++ standards stay binary compatible to previous versions, compiler and standard library vendors put a lot of effort into achieving it anyway. However, there are a very few breaking changes over the years and enforcing a C++ standard on a CMake target can make it incompatible. The C++ language itself is backward compatible, though. Hence, there is no need to lock a project at all costs to a certain standard version. Projects can pass down variables like `CMAKE_CXX_STANDARD` to third-party CMake projects to build and link everything in a guaranteed binary compatible way. Tl;dr, when the variable is set, do not override it manually at target-level.

The alternative is to use `CMAKE_CXX_STANDARD` in this project as well and when it is passed as command-line argument, it simply will be overriden as specified by CMake's rules: 

```suggestion
set(CMAKE_CXX_STANDARD 98)
set(CMAKE_CXX_EXTENSIONS OFF) # When you are concerned about standards, you usually really want the standard
```

---

Unconditionally setting C++ standard will cause conflicts.

---

> Unconditionally setting C++ standard will cause conflicts.

Exact, this is the whole point of my comment. Sorry if I didn't communicate that clearly. My first suggestion is an explicit solution, my alternative suggestion may not be that obvious as the CMake command-line overriding rules are more implicit (read: it is not an unconditional set). But of course you could add that as cache variable as well to make it more explicit (maybe marked as advanced, as many users won't have an issue at all in more simple project structures). The first suggestion results in the least additional CMake code while still open the possibility for users to control binary compatibility through cache variables / command-line arguments.

---

Just to be clear: target_compile_features(imgui PRIVATE cxx_std_98)

doesn't set the compiler into c++98 mode. It tells cmake to set it **at least** into c++98. If you e.g. set Cmake_CXX_STANDARD to 11,14,17 ... cmake will still use that standard.

I honestly don't see the point of all this though. My Guess is that pretty much any compiler supported by cmake will anyway default to c++98 or newer (actually I don't know if any compiler offers any form of pre c++98 mode).

---

> doesn't set the compiler into c++98 mode. It tells cmake to set it at least into c++98. If you e.g. set Cmake_CXX_STANDARD to 11,14,17 ... cmake will still use that standard.

TIL. Thank you 👍🏻

---

@Mike-Devel is correct here. :) Resolving.